### PR TITLE
Nithin/problem

### PR DESCRIPTION
**Pull request by nithin-mn**

### CODE REVIEW

File: AndroidProblem/.gitignore [hunk 1]
```diff
@@ -0,0 +1,25 @@
+*.iml
+.gradle
+/local.properties
+/.idea/caches
+/.idea/libraries
+/.idea/modules.xml
+/.idea/workspace.xml
+/.idea/navEditor.xml
+/.idea/assetWizardSettings.xml
```
Comment (sachin-surya-soft) [inline AndroidProblem/.gitignore]: You have already added `.idea/` below, so no need for all these.

File: AndroidProblem/.gitignore [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+.DS_Store
+/build
```
Comment (sachin-surya-soft) [inline AndroidProblem/.gitignore:4]: already added this?

File: AndroidProblem/app/src/main/java/com/example/android/androidproblem/CustomAdapter.kt [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+import android.view.ViewGroup
+import android.widget.ArrayAdapter
+import android.widget.ImageView
+import android.widget.TextView
+import coil.ImageLoader
+import coil.load
+import coil.request.ImageRequest
+
+class CustomAdapter(
+    private val context: Activity,
```
Comment (sachin-surya-soft) [inline AndroidProblem/app/src/main/java/com/example/android/androidproblem/CustomAdapter.kt]: Isn't this should be of type `Context`?

File: AndroidProblem/app/src/main/java/com/example/android/androidproblem/CustomAdapter.kt [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    private val description: ArrayList<String>,
+    private val imageurl: ArrayList<String>
```
Comment (sachin-surya-soft) [inline AndroidProblem/app/src/main/java/com/example/android/androidproblem/CustomAdapter.kt]: Why? You can say `ArrayList<UserInfo>`?

File: AndroidProblem/app/src/main/java/com/example/android/androidproblem/MainActivity.kt [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+class MainActivity : AppCompatActivity() {
+    val emailId by lazy { findViewById<EditText>(R.id.email_id) }
+    val userList by lazy { findViewById<ListView>(R.id.user_list) }
+    val done by lazy { findViewById<Button>(R.id.done_button) }
+    val userString = ArrayList<String>()
+    val listOfImageUrl = ArrayList<String>()
+    val apiService by lazy { RestApiService() }
+    val gson = Gson()
+    lateinit var sharedPref:SharedPreferences
```
Comment (sachin-surya-soft) [inline AndroidProblem/app/src/main/java/com/example/android/androidproblem/MainActivity.kt]: All these can be private?

File: AndroidProblem/app/src/main/java/com/example/android/androidproblem/MainActivity.kt [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_main)
+        GlobalScope.launch{
+            getCache()
+        }
+        done.setOnClickListener {
+            if (emailId.text.isNullOrEmpty()) {
+                Log.d(TAG, "invalid email")
```
Comment (sachin-surya-soft) [inline AndroidProblem/app/src/main/java/com/example/android/androidproblem/MainActivity.kt]: This log is not visible to user? We need to tell user that you need to type something in email field.
May be use a toast?

File: AndroidProblem/app/src/main/java/com/example/android/androidproblem/MainActivity.kt [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+            listOfImageUrl.add(it.imageUrl)
+        }
+        createListView(userString, listOfImageUrl)
+    }
+
+    private fun updateCache() {
+        GlobalScope.launch {
+            sharedPref.edit().putString("user_string", gson.toJson(userString)).apply()
+            sharedPref.edit().putString("url_string", gson.toJson(listOfImageUrl)).apply()
+        }
```
Comment (sachin-surya-soft) [inline AndroidProblem/app/src/main/java/com/example/android/androidproblem/MainActivity.kt]: I am not sure you need to do this in background?

File: AndroidProblem/app/src/main/java/com/example/android/androidproblem/UserInfo.kt [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+data class User(
+        @SerializedName("emailId") val emailId: String,
+        @SerializedName("imageUrl") val imageUrl: String,
+        @SerializedName("firstName") val firstName: String,
+        @SerializedName("lastName") val lastName: String
+){
+        override fun toString(): String {
+                return "$firstName $lastName\n$emailId\n"
+        }
+}
```
Comment (sachin-surya-soft) [inline AndroidProblem/app/src/main/java/com/example/android/androidproblem/UserInfo.kt]: Why do you need this?

File: AndroidProblem/app/src/main/res/layout/user_item.xml [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+    <ImageView
+        android:id="@+id/user_image"
+        android:layout_width="150dp"
+        android:layout_height="200dp"
+        android:layout_marginStart="8dp"
+        android:contentDescription="@string/displayPicture"
+        android:padding="16dp"
+        app:layout_constraintBottom_toBottomOf="parent"
+        app:layout_constraintEnd_toStartOf="@+id/user_detail"
+        app:layout_constraintHorizontal_bias="0.5"
```
Comment (sachin-surya-soft) [inline AndroidProblem/app/src/main/res/layout/user_item.xml]: Why?

File: AndroidProblem/app/src/main/res/layout/user_item.xml [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+        android:id="@+id/user_detail"
+        android:layout_width="0dp"
+        android:layout_height="wrap_content"
+        android:layout_marginStart="16dp"
+        android:layout_marginTop="8dp"
+        android:layout_marginEnd="8dp"
+        android:textColor="@android:color/black"
+        app:layout_constraintBottom_toBottomOf="parent"
+        app:layout_constraintEnd_toEndOf="parent"
+        app:layout_constraintHorizontal_bias="0.5"
```
Comment (sachin-surya-soft) [inline AndroidProblem/app/src/main/res/layout/user_item.xml]: Why?

**Review by sachin-surya-soft — CHANGES_REQUESTED** [review]

File: AndroidProblem/app/src/main/java/com/example/android/androidproblem/CustomAdapter.kt [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+import coil.load
+import coil.request.ImageRequest
+
+class CustomAdapter(
+    private val contex: Context,
+    private val description: List<User>,
+) : ArrayAdapter<User>(contex, R.layout.user_item ,description) {
+
+    override fun getView(position: Int, view: View?, parent: ViewGroup): View {
+        val inflater=LayoutInflater.from(contex)
```
Comment (sachin-surya-soft) [inline AndroidProblem/app/src/main/java/com/example/android/androidproblem/CustomAdapter.kt]: Formatting?

File: AndroidProblem/app/src/main/java/com/example/android/androidproblem/MainActivity.kt [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+        if (userString != null) {
+            createListView(userJson)
+        }
+    }
+
+    private fun getHttpResponse() {
+        val userInfo = UserRequest(emailId.text.toString())
+        apiService.addUser(userInfo) {
+            if (it?.items != null) {
+                Log.d(TAG, "got response")
```
Comment (sachin-surya-soft) [inline AndroidProblem/app/src/main/java/com/example/android/androidproblem/MainActivity.kt]: Not required?

File: AndroidProblem/app/src/main/java/com/example/android/androidproblem/MainActivity.kt [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+        val userJson = gson.fromJson<ArrayList<User>>(userString, type)
+        if (userString != null) {
+            createListView(userJson)
+        }
+    }
+
+    private fun getHttpResponse() {
+        val userInfo = UserRequest(emailId.text.toString())
+        apiService.addUser(userInfo) {
+            if (it?.items != null) {
```
Comment (sachin-surya-soft) [inline AndroidProblem/app/src/main/java/com/example/android/androidproblem/MainActivity.kt]: You can use `let` scope method here?

**Review by sachin-surya-soft — CHANGES_REQUESTED** [review]

File: AndroidProblem/app/src/main/java/com/example/android/androidproblem/MainActivity.kt (hunk 13 above)
Comment (nithin-mn) [reply AndroidProblem/app/src/main/java/com/example/android/androidproblem/MainActivity.kt]: `apiService.addUser(userInfo) {`
            `it?.items.let {`
                `if (it != null) {`
                    `createListView(it)`
                    `sharedPref.edit().putString("user_string", gson.toJson(it))`
                       ` .apply()`
               ` }`
            `}`
       ` }`
Is it correct?

Comment (sachin-surya-soft) [reply AndroidProblem/app/src/main/java/com/example/android/androidproblem/MainActivity.kt]: Yes

File: AndroidProblem/app/src/main/java/com/example/android/androidproblem/CustomAdapter.kt [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+    override fun getView(position: Int, view: View?, parent: ViewGroup): View {
+        val inflater = LayoutInflater.from(contex)
+        val rowView = inflater.inflate(R.layout.user_item, null, true)
+
+        val userDetail: TextView = rowView.findViewById(R.id.user_detail)
+        val userImage: ImageView = rowView.findViewById(R.id.user_image)
+        val str =
+            "${description[position].firstName} ${description[position].lastName}\n${description[position].emailId}"
+        userDetail.text = str
+
```
Comment (sachin-surya-soft) [inline AndroidProblem/app/src/main/java/com/example/android/androidproblem/CustomAdapter.kt]: You can eliminate `str`?

File: AndroidProblem/app/src/main/java/com/example/android/androidproblem/CustomAdapter.kt [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+import coil.request.ImageRequest
+
+class CustomAdapter(
+    private val contex: Context,
+    private val description: List<User>,
+) : ArrayAdapter<User>(contex, R.layout.user_item, description) {
+
+    override fun getView(position: Int, view: View?, parent: ViewGroup): View {
+        val inflater = LayoutInflater.from(contex)
+        val rowView = inflater.inflate(R.layout.user_item, null, true)
```
Comment (sachin-surya-soft) [inline AndroidProblem/app/src/main/java/com/example/android/androidproblem/CustomAdapter.kt]: You can combine these two?

File: AndroidProblem/app/src/main/java/com/example/android/androidproblem/CustomAdapter.kt [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+                onSuccess = { result ->
+                    userImage.setImageDrawable(result)
+                },
+                onError = {
+                    userImage.load(R.drawable.default_image)
+                }
+            )
+            .build()
+
+        val imageLoader = ImageLoader(context)
```
Comment (sachin-surya-soft) [inline AndroidProblem/app/src/main/java/com/example/android/androidproblem/CustomAdapter.kt]: Why do you want to initialize every time a row is created?

File: AndroidProblem/app/src/main/java/com/example/android/androidproblem/MainActivity.kt [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+    private lateinit var sharedPref: SharedPreferences
+    private val type by lazy { object : TypeToken<ArrayList<User>>() {}.type }
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_main)
+        GlobalScope.launch { getCache() }
+        done.setOnClickListener {
+            if (emailId.text.isNullOrEmpty()) {
+                Toast.makeText(this, "Enter valid email id", Toast.LENGTH_SHORT).show()
```
Comment (sachin-surya-soft) [inline AndroidProblem/app/src/main/java/com/example/android/androidproblem/MainActivity.kt]: The message should com from string resources?

File: AndroidProblem/app/src/main/java/com/example/android/androidproblem/MainActivity.kt [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+            } else {
+                getHttpResponse()
+            }
+        }
+    }
+
+    private fun getCache() {
+        sharedPref = getSharedPreferences("myPref", MODE_PRIVATE)
+        val userString = sharedPref.getString("user_string", null)
```
Comment (sachin-surya-soft) [inline AndroidProblem/app/src/main/java/com/example/android/androidproblem/MainActivity.kt]: You can store the key here `user_string` in a `private const`?

File: AndroidProblem/app/src/main/java/com/example/android/androidproblem/MainActivity.kt [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
+        if (userString != null) {
```
Comment (sachin-surya-soft) [inline AndroidProblem/app/src/main/java/com/example/android/androidproblem/MainActivity.kt]: You can use `let`?

**Review by sachin-surya-soft — CHANGES_REQUESTED** [review]

File: AndroidProblem/app/src/main/java/com/example/android/androidproblem/CustomAdapter.kt [hunk 20]
```diff
@@ -0,0 +1,45 @@
...
+import android.view.View
+import android.view.ViewGroup
+import android.widget.ArrayAdapter
+import android.widget.ImageView
+import android.widget.TextView
+import coil.ImageLoader
+import coil.load
+import coil.request.ImageRequest
+
+class CustomAdapter(
```
Comment (sachin-surya-soft) [inline AndroidProblem/app/src/main/java/com/example/android/androidproblem/CustomAdapter.kt]: You name this adapter better? May be `UserListAdapter`?

File: AndroidProblem/app/src/main/java/com/example/android/androidproblem/CustomAdapter.kt [hunk 21]
```diff
@@ -0,0 +1,45 @@
...
+    private val contex: Context,
+    private val description: List<User>,
+) : ArrayAdapter<User>(contex, R.layout.user_item, description) {
+    val imageLoader = ImageLoader(context)
+    val rowView = LayoutInflater.from(contex).inflate(R.layout.user_item, null, true)
+    val userDetail: TextView = rowView.findViewById(R.id.user_detail)
+    val userImage: ImageView = rowView.findViewById(R.id.user_image)
```
Comment (sachin-surya-soft) [inline AndroidProblem/app/src/main/java/com/example/android/androidproblem/CustomAdapter.kt]: May be you want to make these all private?

File: AndroidProblem/app/src/main/java/com/example/android/androidproblem/MainActivity.kt [hunk 22]
```diff
@@ -0,0 +1,65 @@
...
+        }
+    }
+
+    private fun getCache() {
+        sharedPref = getSharedPreferences(name, MODE_PRIVATE)
+        sharedPref.let {
+            val userString = it.getString(key, null)
+            if (userString != null) {
+                createListView(gson.fromJson<ArrayList<User>>(userString, type))
+            }
```
Comment (sachin-surya-soft) [inline AndroidProblem/app/src/main/java/com/example/android/androidproblem/MainActivity.kt]: No I asked you to use `let` in place of `if (userString != null)`, not with sharedPref.
Why do we use `let`?

**Review by sachin-surya-soft — CHANGES_REQUESTED** [review]

File: AndroidProblem/app/src/main/java/com/example/android/androidproblem/MainActivity.kt (hunk 22 above)
Comment (nithin-mn) [reply AndroidProblem/app/src/main/java/com/example/android/androidproblem/MainActivity.kt]: To check nullable types and execute the `let` block only if the value is non-null.

File: AndroidProblem/app/src/main/java/com/example/android/androidproblem/MainActivity.kt [hunk 23]
```diff
@@ -0,0 +1,63 @@
...
+        }
+    }
+
+    private fun createListView(userString: List<User>) {
+        userList.adapter = UserListAdapter(this, userString)
+    }
+}
+
+private const val name = "myPref"
+private const val key = "user_string"
```
Comment (sachin-surya-soft) [inline AndroidProblem/app/src/main/java/com/example/android/androidproblem/MainActivity.kt]: Always keep the name of `private const val` in capitals.

**Review by sachin-surya-soft — CHANGES_REQUESTED** [review]

**Review by sachin-surya-soft — APPROVED** [review]

**Pull request merged**